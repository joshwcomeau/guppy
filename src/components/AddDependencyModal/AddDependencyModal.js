// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import { InstantSearch, SearchBox, Hits } from 'react-instantsearch/dom';

import Modal from '../Modal';
import ModalHeader from '../ModalHeader';
import AddDependencySearchBox from '../AddDependencySearchBox';
import AddDependencySearchResult from '../AddDependencySearchResult';
import Spacer from '../Spacer';

type Props = {
  isVisible: boolean,
  onDismiss: () => void,
};

type State = {
  searchTerm: string,
};

class AddDependencyModal extends Component<Props, State> {
  state = {
    searchTerm: '',
  };

  updateSearchTerm = (ev: SyntheticInputEvent<*>) => {
    this.setState({ searchTerm: ev.target.value });
  };

  render() {
    const { isVisible, onDismiss } = this.props;
    const { searchTerm } = this.state;

    return (
      <InstantSearch
        appId="OFCNCOG2CU"
        apiKey="f54e21fa3a2a0160595bb058179bfb1e"
        indexName="npm-search"
      >
        <Modal
          width={620}
          height={800}
          isVisible={isVisible}
          onDismiss={onDismiss}
        >
          <Wrapper>
            <ModalHeader title="Add New Dependency" theme="blueish">
              <AddDependencySearchBox />
            </ModalHeader>

            <HitsWrapper>
              <Hits hitComponent={AddDependencySearchResult} />
            </HitsWrapper>
            <Spacer size={25} />
          </Wrapper>
        </Modal>
      </InstantSearch>
    );
  }
}

const Wrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
`;

const HitsWrapper = styled.div`
  flex: 1;
  padding: 25px;
  padding-top: 0;
  overflow: auto;
`;

export default AddDependencyModal;
