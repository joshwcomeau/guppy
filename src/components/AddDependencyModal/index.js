// @flow
export { default } from './AddDependencyModal';
