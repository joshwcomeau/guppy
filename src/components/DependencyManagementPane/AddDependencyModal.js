// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import { InfiniteHits } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';

import Modal from '../Modal';
import ModalHeader from '../ModalHeader';
import Spacer from '../Spacer';
import FadeOnChange from '../FadeOnChange';
import AddDependencySearchBox from './AddDependencySearchBox';
import AddDependencySearchResult from './AddDependencySearchResult';
import AddDependencyInitialScreen from './AddDependencyInitialScreen';

type Props = {
  isVisible: boolean,
  onDismiss: () => void,
  // Provided by `connectStateResults`:
  searchResults: {
    query?: string,
  },
};

export class AddDependencyModal extends Component<Props> {
  render() {
    const { isVisible, onDismiss, searchResults } = this.props;

    const hasSearchQuery = searchResults && !!searchResults.query;

    return (
      <Modal
        width={620}
        height={800}
        isVisible={isVisible}
        onDismiss={onDismiss}
      >
        <Wrapper>
          <ModalHeader title="Add New Dependency" theme="blueish">
            <AddDependencySearchBox />
          </ModalHeader>

          <HitsWrapper>
            <FadeOnChange changeKey={!!hasSearchQuery}>
              {hasSearchQuery ? (
                <InfiniteHits hitComponent={AddDependencySearchResult} />
              ) : (
                <AddDependencyInitialScreen />
              )}
            </FadeOnChange>
          </HitsWrapper>
          <Spacer size={25} />
        </Wrapper>
      </Modal>
    );
  }
}

const Wrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
`;

const HitsWrapper = styled.div`
  flex: 1;
  padding: 25px;
  padding-top: 0;
  overflow: auto;
`;

export default connectStateResults(AddDependencyModal);
