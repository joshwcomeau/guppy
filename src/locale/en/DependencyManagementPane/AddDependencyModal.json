{
  "dependencyManagementPane.addDependencyModal.main.title": "Add New Dependency"
}