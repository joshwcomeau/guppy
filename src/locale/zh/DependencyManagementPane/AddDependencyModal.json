{
  "dependencyManagementPane.addDependencyModal.main.title": "添加新的依赖"
}