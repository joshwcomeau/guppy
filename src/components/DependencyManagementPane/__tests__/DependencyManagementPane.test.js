import React from 'react';
import { shallow, mount } from 'enzyme';

import {
  DependencyManagementPane,
  DependencyButton,
  Dependencies,
  MainContent,
  AddDependencyButton,
} from '../DependencyManagementPane';
import AddDependencyModal from '../AddDependencyModal';

describe('DependencyManagement component', () => {
  let wrapper;
  let instance;

  const project = {
    id: 'projectId',
    dependencies: [
      {
        name: 'React',
        status: 'idle',
        location: 'dependencies',
        version: '16.0.0',
      },
      {
        name: 'Redux',
        status: 'idle',
        location: 'devDependencies',
        version: '4.0.0',
      },
    ],
  };

  beforeEach(() => {
    wrapper = shallow(<DependencyManagementPane project={project} />);
    instance = wrapper.instance();
  });

  describe('Rendering', () => {
    it('should render a dependency list with two buttons', () => {
      expect(wrapper.find(Dependencies)).toMatchSnapshot();
    });

    it('should render main content for selected dependency (default)', () => {
      expect(wrapper.find(MainContent)).toMatchSnapshot();
    });

    it('should render main content for selected dependency (installing)', () => {
      wrapper.setProps({
        project: {
          ...project,
          dependencies: [
            {
              ...project.dependencies[0],
              status: 'installing',
            },
            project.dependencies[1],
          ],
        },
      });

      expect(wrapper.find(MainContent)).toMatchSnapshot();
    });

    it('should render list addon for React dependency (not selected)', () => {
      expect(
        mount(instance.renderListAddon(project.dependencies[0], false))
      ).toMatchSnapshot();
    });

    it('should render list addon for React dependency (selected)', () => {
      expect(
        mount(instance.renderListAddon(project.dependencies[0], true))
      ).toMatchSnapshot();
    });
  });

  describe('Component logic', () => {
    it('should select dependency on click', () => {
      // We're selecting the last dep. so we're having an index larger than 0
      const button = wrapper.find(DependencyButton).last();
      button.simulate('click');
      expect(instance.state.selectedDependencyIndex).toBe(
        project.dependencies.length - 1
      );
    });

    it('should select the newly added dependency', () => {
      const updatedProject = {
        project: {
          ...project,
          dependencies: [
            ...project.dependencies,
            {
              name: 'Loadash',
              status: 'idle',
              location: 'dependencies',
            },
          ],
        },
      };
      wrapper.setProps(updatedProject);

      expect(instance.state.selectedDependencyIndex).toBe(
        updatedProject.project.dependencies.length - 1
      );
    });

    it('should change selection if selectedIndex out of range', () => {
      wrapper.setState({
        selectedDependencyIndex: 1,
      });
      wrapper.setProps({
        project: {
          ...project,
          dependencies: [project.dependencies[0]],
        },
      });

      expect(instance.state.selectedDependencyIndex).toBe(0);
    });

    it('should open addingNewDependency modal', () => {
      const button = wrapper.find(AddDependencyButton);
      button.simulate('click');

      expect(instance.state.addingNewDependency).toBeTruthy();
    });

    it('should hide addingNewDependency modal on dismiss', () => {
      wrapper.setState({
        addingNewDependency: true,
      });
      const modal = wrapper.find(AddDependencyModal);

      modal.prop('onDismiss')();
      expect(instance.state.addingNewDependency).toBeFalsy();
    });
  });
});
