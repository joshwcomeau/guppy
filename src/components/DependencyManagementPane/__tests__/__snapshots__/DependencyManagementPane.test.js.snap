// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyManagement component Rendering should render a dependency list with two buttons 1`] = `
<styled.div>
  <styled.button
    isSelected={true}
    key="React"
    onClick={[Function]}
  >
    <styled.span>
      React
    </styled.span>
    <styled.span
      isSelected={true}
    >
      16.0.0
    </styled.span>
  </styled.button>
  <styled.button
    isSelected={false}
    key="Redux"
    onClick={[Function]}
  >
    <styled.span>
      Redux
    </styled.span>
    <styled.span
      isSelected={false}
    >
      4.0.0
    </styled.span>
  </styled.button>
</styled.div>
`;

exports[`DependencyManagement component Rendering should render list addon for React dependency (not selected) 1`] = `
.c0 {
  padding-left: 10px;
  font-size: 16px;
  color: rgba(0,0,0,0.66);
}

<styled.span
  isSelected={false}
>
  <span
    className="c0"
  >
    16.0.0
  </span>
</styled.span>
`;

exports[`DependencyManagement component Rendering should render list addon for React dependency (selected) 1`] = `
.c0 {
  padding-left: 10px;
  font-size: 16px;
  color: rgba(255,255,255,0.66);
}

<styled.span
  isSelected={true}
>
  <span
    className="c0"
  >
    16.0.0
  </span>
</styled.span>
`;

exports[`DependencyManagement component Rendering should render main content for selected dependency (default) 1`] = `
<styled.div>
  <DependencyDetails
    dependency={
      Object {
        "location": "dependencies",
        "name": "React",
        "status": "idle",
        "version": "16.0.0",
      }
    }
    projectId="projectId"
  />
</styled.div>
`;

exports[`DependencyManagement component Rendering should render main content for selected dependency (installing) 1`] = `
<styled.div>
  <DependencyInstalling
    name="React"
    queued={false}
  />
</styled.div>
`;
